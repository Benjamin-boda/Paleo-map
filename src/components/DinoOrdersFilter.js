import React, {useState, useEffect} from "react";
import {Form} from "react-bootstrap";

export const DinoOrdersFilter = ({loaded, dinosauriaOrderArray, setDinosauria}) => {
    const [order, setOrder] = useState("")

    const onOrderChange = (e) => {
        setOrder(e.target.value)
    }

    useEffect(() => {
        if (order !== "") {
            const NoUndefined = dinosauriaOrderArray.filter((dino) => dino.cll !== undefined)
            const orderFiltered = NoUndefined.filter((dino) => dino.cll === order)
            setDinosauria(orderFiltered)
        } else {
            setDinosauria(dinosauriaOrderArray)
        }
         
    }, [order, loaded])

    return (
        
            <Form.Control className="filters__select" as="select" onChange={onOrderChange}>
                <option value="">--Filter by order--</option>
                <option value="Saurischia">Saurischia</option>
                <option value="Ornithischia">Ornithischia</option>
                <option value="Aves">Aves</option>
            </Form.Control>
        
    )
}