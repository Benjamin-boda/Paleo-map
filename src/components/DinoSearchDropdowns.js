import React, {useEffect, useState} from "react";
import {Dropdown} from "react-bootstrap";

export const DinoSearchDropdowns = ({dinosauria, setOpenDropdown, keyword, setIdModal, setModalIsOpen}) => {
    const [dinoSearch, setdinoSearch] = useState([])

    useEffect(() => {
        const NoUndefined = dinosauria.filter((dino) => dino.gnl !== undefined)
        const NoOtherProperties = NoUndefined.map(({gnl, oid, cll}) => ({gnl, oid, cll}))
        const NoDuplicate = NoOtherProperties.filter((dino, index, self) =>
            index === self.findIndex((element) => (
                element["gnl"] === dino["gnl"]
            ))
        )
        const search = NoDuplicate.filter((dino) => dino.gnl.toLowerCase().indexOf(keyword) > -1)
        setdinoSearch(search.slice(0,10))
        
        if (keyword === "") {
            setOpenDropdown(false)
        }
    }, [keyword])

    return (
        <div>
            {dinoSearch.map((dino) => 
                <li className="header__dropdown" key={dino.oid} 
                    onClick={() => {
                        setModalIsOpen(true)
                        setIdModal(dino.oid)
                        setOpenDropdown(false)
                }}>
                    {dino.gnl} - {dino.cll}
                </li>
            )}
            <Dropdown.Divider />
            <li className="header__dropdown" onClick={() => setOpenDropdown(false)} >Close</li>
        </div>
    )
}