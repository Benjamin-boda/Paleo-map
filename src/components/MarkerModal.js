import React, {useState, useEffect} from "react";
import {Modal, Button} from "react-bootstrap";

export const MarkerModal = ({dino, modalIsOpen, setModalIsOpen}) => {
    const [wikiInfos, setWikiInfos] = useState([])
    const [sourceError, setSourceError] = useState(false)
    

    useEffect(() => {
        async function getWikiApi() {
            await fetch(`https://en.wikipedia.org/w/api.php?origin=*&action=query&prop=pageimages&titles=${dino.gnl}&pithumbsize=500&format=json`)
                .then((req) => req.json())
                .then((res) => {
                    const wikiSource = res.query.pages
                    setWikiInfos(Object.values(wikiSource).map(({ thumbnail }) => thumbnail)[0].source)
                    setSourceError(false)
                })
                .catch((err) => setSourceError(true))
        }

        getWikiApi()
    }, [])

    return (
        <Modal aria-labelledby="contained-modal-title-vcenter" centered show={modalIsOpen} onHide={() => setModalIsOpen(!modalIsOpen)}>
            <Modal.Header className="modal__title">{dino.gnl}</Modal.Header>
            <Modal.Body>
                {sourceError ? undefined : <img className="modal__image" src={wikiInfos}/>}
                <p className="modal__text"><span className="modal__text__bold">Location : </span>{dino.ggc}</p>
                <p className="modal__text__bold">Clade : <a className="modal__link" target="_blank" href={`https://en.wikipedia.org/wiki/${dino.cll}`}>{dino.cll}</a></p>
                <p className="modal__text__bold">Details on Wikipedia : <a className="modal__link" target="_blank" href={`https://en.wikipedia.org/wiki/${dino.gnl}`}>{dino.gnl}</a></p>
            </Modal.Body>
            <Modal.Footer>
                <Button variant="secondary" size="sm" onClick={() => setModalIsOpen(!modalIsOpen)}>Close</Button>
            </Modal.Footer>
        </Modal>
    )
}