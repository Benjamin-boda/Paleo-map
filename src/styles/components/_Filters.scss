.filters__background {
    background: $light-grey;
    height: 15vh;
}

.filters {
    align-items: flex-end;
    display: flex;
    justify-content: space-between;
    margin: 0 auto;
    max-width: 110rem;
}

.filters__select {
    font-family: "OpenSansRegular";
    font-size: 1rem;
    margin: $s-size 0;
    width: 40%;

    @media (min-width: $desktop-breakpoint) {
        font-size: $font-size-xsmall;
        margin: $l-size $m-size 0 $m-size;
        width: 40%;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin: $s-size $m-size;
        width: 30%;
    }
}