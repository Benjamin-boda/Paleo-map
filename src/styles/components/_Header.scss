.header {
    background: $dark-grey;
}

.header__content {
    align-items: center;
    display: flex;
    justify-content: flex-start;
    padding: $s-size 0;
    margin: 0 auto;
    max-width: 110rem;

    @media (min-width: $desktop-breakpoint) {
        justify-content: space-between;
        margin: 0 $m-size;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin: 0 auto;
    }
}

.header__logo {
    filter: invert(100%) sepia(11%) saturate(0%) hue-rotate(83deg) brightness(105%) contrast(101%);
    height: 3rem;
    margin-right: $m-size;
    width: 3rem;

    @media (min-width: $desktop-breakpoint) {
        height: 4rem;
        margin-right: 0;
        width: 4rem;
    }
}

.header__text {
    color: white;
    font-family: "OpenSansExtraBold";
    font-size: $font-size-small;
    margin-right: $m-size;

    @media (min-width: $desktop-breakpoint) {
        font-size: $font-size-medium;
        margin-right: 0;
    }
}

.header__input {
    width: 50%;

    @media (min-width: $desktop-breakpoint) {
        width: 40%;
    }
}

.dropdown {
    position: absolute;
}

.header__dropdown__content {
    background: white;
    border-radius: 0 0 0 1rem;
    padding: 0 $m-size $m-size $m-size;
    position: relative;
    

    @media (min-width: $desktop-breakpoint) {
        left: 13rem;
        top: 3.1rem;
    }
}

.header__dropdown {
    font-family: "OpenSansRegular";
    font-size: $font-size-xsmall;
    list-style-type: none;
    text-decoration: none;

    @media (min-width: $desktop-breakpoint-large) {
        font-size: $font-size-small;
    }
}

.header__dropdown:hover {
    cursor: pointer;
    background: grey;
}

.InputGroup__Text {
    background: $light-grey;
    color: black;
    font-family: "OpenSansLight";
}