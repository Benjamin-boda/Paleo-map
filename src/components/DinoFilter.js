import React, { useEffect, useState } from "react";
import {countryList} from "../map/country";
import {DinoOrdersFilter} from "./DinoOrdersFilter";
import {Form} from "react-bootstrap";

export const DinoFilter = ({loaded, continentOrCountry, setContinentOrCountry, dinosauriaOrderArray, setDinosauria}) => {
    const [continentName, setContinentName] = useState(null)
    const [countryFiltered, setCountryFiltered] = useState(countryList)

    useEffect(() => {
        if (continentName === null || continentOrCountry === "") {
            setCountryFiltered(countryList)
        } else {
            setCountryFiltered(countryList.filter((country) => country.Continent_Name === continentName)) 
        }
       
    }, [continentName])
    

    

    const onContinentChange = (e) => {
        setContinentOrCountry(e.target.value)
        setContinentName(e.target.selectedOptions[0].outerText)
    }

    const onCountryChange = (e) => {
        setContinentOrCountry(e.target.value)
    }

    return (
        <div className="filters__background">
            <div className="filters">

                <Form.Control id="continent" className="filters__select" as="select" onChange={onContinentChange}>
                    <option value="">--Filter by continent--</option>
                    <option value="&cc=ATA">Antarctica</option>
                    <option value="&cc=AFR">Africa</option>
                    <option value="&cc=ASI">Asia</option>
                    <option value="&cc=EUR">Europe</option>
                    <option value="&cc=NOA">North America</option>
                    <option value="&cc=OCE">Oceania</option>
                    <option value="&cc=SOA">South America</option>
                </Form.Control>

                <Form.Control className="filters__select" as="select" onChange={onCountryChange}>
                    <option value="">--Filter by country--</option>
                    {countryFiltered.map((country) => 
                        <option key={country.Country_Name + country.Continent_Code} value={`&cc=${country.Two_Letter_Country_Code}`}>
                            {country.Country_Name.includes(",") ? country.Country_Name.substring(0, country.Country_Name.indexOf(",")) : country.Country_Name}
                        </option>   
                    )}
                </Form.Control>

                <DinoOrdersFilter
                    loaded={loaded}
                    dinosauriaOrderArray={dinosauriaOrderArray} 
                    setDinosauria={setDinosauria}
                />
            </div>
        </div>
        
    )
}