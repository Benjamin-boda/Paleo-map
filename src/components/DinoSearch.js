import React, {useState} from "react";
import {InputGroup, FormControl, Dropdown} from "react-bootstrap";
import {DinoSearchDropdowns} from "./DinoSearchDropdowns";

export const DinoSearch = ({dinosauria, loaded, setIdModal, setModalIsOpen}) => {
    const [keyword, setKeyword] = useState("")
    const [openDropdown, setOpenDropdown] = useState(false)

    const onKeywordChange = (e) => {
        const keywordLower = e.target.value
        setKeyword(keywordLower.toLowerCase())
        setOpenDropdown(true)
    }

    return (
        <div className="header">
            <div className="header__content">

                <img className="header__logo" src={require("../images/logo.png").default}/>

                <h1 className="header__text">PALEO MAP</h1>

                <InputGroup 
                    size="lg"
                    className="header__input" 
                    onChange={onKeywordChange}
                    type="text"
                    placeholder="Search a specimen..."
                    value={keyword}
                >
                        <InputGroup.Prepend>
                            <InputGroup.Text className="InputGroup__Text" id="inputGroup-sizing-lg">Search a specimen</InputGroup.Text>
                        </InputGroup.Prepend>
                        <FormControl
                        aria-label="Large"
                        aria-describedby="inputGroup-sizing-sm"
                        />

                        <Dropdown className="dropdown">
                            {openDropdown && loaded ? 
                                <ul className="header__dropdown__content">
                                    <DinoSearchDropdowns 
                                        
                                        dinosauria={dinosauria} 
                                        setOpenDropdown={setOpenDropdown} 
                                        keyword={keyword}
                                        setIdModal={setIdModal} 
                                        setModalIsOpen={setModalIsOpen}
                                    />
                                </ul> : undefined}
                        </Dropdown>
                </InputGroup>

            </div>
        </div>
    ) 
}