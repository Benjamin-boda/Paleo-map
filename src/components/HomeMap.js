import React, {useState, useEffect} from "react";
import { ComposableMap, Geographies, Geography, Marker, ZoomableGroup} from "react-simple-maps";
import world from '../map/world.json';
import {MarkerModal} from "./MarkerModal";
import {DinoFilter} from "./DinoFilter";
import {DinoSearch} from "./DinoSearch";

export const HomeMap = () => {
    const [dinosauria, setDinosauria] = useState([])
    const [idModal, setIdModal] = useState(0)
    const [modalIsOpen, setModalIsOpen] = useState(false)
    const [continentOrCountry, setContinentOrCountry] = useState([])
    const [loaded, setLoaded] = useState(false)
    const [dinosauriaOrderArray, setDinosauriaOrderArray] = useState([])
    const [textId, setTextId] = useState("")

    useEffect(() => {
        async function getPBDBApi() {
            setLoaded(false)
            await fetch(`https://paleobiodb.org/data1.2/specs/list.json?datainfo&rowcount&base_name=Dinosauria&idqual=certain${continentOrCountry}&show=class,abund,coll,coords,loc,geo,acconly&textresult`)
                .then((req) => req.json())
                .then((res) => {
                    setDinosauria(res.records)
                    setDinosauriaOrderArray(res.records)
                    setLoaded(true)
                })
        }

        getPBDBApi()
    }, [continentOrCountry])

    return (
        <div>
            <DinoSearch loaded={loaded} dinosauria={dinosauria} setIdModal={setIdModal} setModalIsOpen={setModalIsOpen}/>

            <div className="map">
                <ComposableMap width={window.innerWidth} height={window.innerHeight * 0.8}>
                    <ZoomableGroup zoom={1} center={[0,0]}>
                        <Geographies geography={world}>
                            {({ geographies }) =>
                                geographies.map(geo => <Geography key={geo.rsmKey} geography={geo} fill="#000000" stroke="#FFF"/>)
                            }
                        </Geographies>
                        {dinosauria.map((dino) =>
                            
                            
                            <Marker key={dino.oid} coordinates={[dino.lng, dino.lat]} 
                                onMouseEnter={() => setTextId(dino.oid)}
                                onMouseLeave={() => setTextId("")}
                                style={{
                                    default: { fill: "#666" },
                                    hover:   { fill: "#ff4500" },
                                    pressed: { fill: "#fff44f" },
                                }}>
                                
                                <circle 
                                    onClick={() => {
                                        setModalIsOpen(!modalIsOpen)
                                        setIdModal(dino.oid)}}
                                        cx="12" cy="10" r="3"  
                                />
                                {textId === dino.oid ? <text className="marker__text" style={{ fontFamily: "OpenSansRegular"}} textAnchor="middle">{dino.gnl}</text> : undefined}
                                {idModal === dino.oid ? <MarkerModal dino={dino} modalIsOpen={modalIsOpen} setModalIsOpen={setModalIsOpen} /> : undefined}
                            </Marker>
                        )}  
                    </ZoomableGroup>   
                </ComposableMap>
            </div>
            
            <DinoFilter
                loaded={loaded}
                continentOrCountry={continentOrCountry}  
                setContinentOrCountry={setContinentOrCountry} 
                dinosauriaOrderArray={dinosauriaOrderArray} 
                setDinosauria={setDinosauria}
            />
        </div>
    )
}