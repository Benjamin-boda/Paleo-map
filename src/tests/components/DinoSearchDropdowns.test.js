import React from "react";
import { DinoSearchDropdowns } from "../../components/DinoSearchDropdowns";
import ReactTestRenderer  from 'react-test-renderer';

let wrapper, tree;

beforeEach(() => {
    wrapper = ReactTestRenderer.create(
        <DinoSearchDropdowns />)
        .toJSON()
});

test("should render dino search dropdowns correctly", () => {
    expect(wrapper).toMatchSnapshot();
});