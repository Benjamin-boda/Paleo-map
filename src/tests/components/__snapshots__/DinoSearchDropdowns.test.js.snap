// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render dino search dropdowns correctly 1`] = `
<div>
  <div
    className="dropdown-divider"
    role="separator"
  />
  <li
    className="header__dropdown"
    onClick={[Function]}
  >
    Close
  </li>
</div>
`;
